src/Raven.Server/Documents/ETL/Providers/OLAP/OlaptEtl.cs

```diff
@@ -144,35 +144,64 @@ protected override IEnumerator<ToOlapItem> ConvertTimeSeriesDeletedRangeEnumerat
         protected override int LoadInternal(IEnumerable<OlapTransformedItems> records, DocumentsOperationContext context, OlapEtlStatsScope scope)
         {
             var count = 0;
-
+            var uploadedFiles = new List<UploadInfo>();
+            var localFiles = new List<string>();
             var outerScope = scope.For(EtlOperations.LoadLocal, start: false);
-
-            foreach (var transformed in records)
+            try
             {
-                outerScope.NumberOfFiles++;
-                string localPath = null;
-                try
+                foreach (var transformed in records)
                 {
-                    string folderName, fileName, safeFolderName;
+                    outerScope.NumberOfFiles++;
+
                     using (outerScope.Start())
                     using (var loadScope = outerScope.For($"{EtlOperations.LoadLocal}/{outerScope.NumberOfFiles}"))
                     {
-                        localPath = transformed.GenerateFile(out folderName, out safeFolderName, out fileName);
+                        string localPath = transformed.GenerateFile(out UploadInfo uploadInfo);
+                        localFiles.Add(localPath);
 
-                        loadScope.FileName = fileName;
+                        loadScope.FileName = uploadInfo.FileName;
                         loadScope.NumberOfFiles = 1;
 
                         count += transformed.Count;
+
+                        if (AnyRemoteDestinations == false) 
+                            continue;
+
+                        UploadToRemoteDestinations(localPath, uploadInfo, scope);
+                        uploadedFiles.Add(uploadInfo);
                     }
+                }
+            }
+            catch (Exception)
+            {
+                // an error occurred during the upload phase.
+                // try delete the uploaded files in order to avoid duplicates.
+                // we will re upload these files in the next batch.
 
-                    if (AnyRemoteDestinations)
-                        UploadToServer(localPath, folderName, fileName, safeFolderName, scope);
+                foreach (var uploadInfo in uploadedFiles)
+                {
+                    TryDeleteFromRemoteDestinations(uploadInfo);
                 }
-                finally
+
+                foreach (var file in localFiles)
```

Comment:
If you have local setting you still will be deleting those?

Reply:
Also, how the local file structure will look like after few failures here?

Reply:
> If you have local setting you still will be deleting those?

Yes, because we will be regenerating them during next run

Reply:
> Also, how the local file structure will look like after few failures here?

This is just a cleanup, in theory we do not need to clean anything since e.g. Athena queries should be handling duplicates anyway

Reply:
That will mean, that if we configured _both_ S3 and local... non of them will make progress if the upload is failing.
I think a nicer approach will be to continue the ETL and generate more local files, and try to upload them later..

Reply:
This is not how ETL works :)